transcom/mymove
# B 18027 - STAGING -- create service layer to create download payload (Part 2: with resolution to correct deployment issue in STAGING)

## [B 18027](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A870406)
### [E-04799](https://www13.v1host.com/USTRANSCOM38/Epic.mvc/Summary?oidToken=Epic%3A864544)

## Summary

For STAGING:

Created new service to merge all user uploaded documents into one PDF file for locator/move code. The service has been designed to generate PDF file for three use cases:

Note: This PR is the second PR for this branch. The original PR was approved and merged into main; however, there was an issue with the deployment to staging. Application was failing to start due to a specific line of code in this branch in paperwork/generator.go. A factory method was attempting to create dir in the temp directory. This threw a READONLY error and thus prevented the application from starting in STAGING. This issue is only present in STAGING due to the usage of AWS cloud storage(S3). S3 currently doesn't have temp directory configured. To resolve this, an in memory file system will be leveraged for PDF generation (generator.go).

1. Merge all order and amendment documents into one file
2. Merge all order docs into one file
3. Merge all amendment docs into one file

This service will be integrated into the [new Prime API](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A868626). [PR](https://github.com/transcom/mymove/pull/11646)

Note:  This PR will also cover the handling of images meant for [B-18092](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A871066) for sprint 3.

 Sample PDF:
[AMDORD-2024-01-10T15-57-09.633Z.pdf](https://github.com/transcom/mymove/files/13896855/AMDORD-2024-01-10T15-57-09.633Z.pdf)

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

Best way to test is to invoke the API via Postman

1. Find locator that is permissible for download and invoke via Postman (see SQL below)
2. Response will display in Postman view as PDF
3. Save response and open in PDF viewer. Check bookmarks for outline of file labeling what is what -- orders and amendments. File should merge everything together --  PDF and images types.

**GET:  /moves/{locator}/order/download**

Business Rule Validations:
- Move owner/client's origin duty location does not provide services counseling 

Find locator code where it has n number of order uploaded documents (change count, by default its set as 1) and meets the business rule validations.
```
SELECT mainSubquery.move_id
	,moves.locator
	,moves.STATUS
	,mainSubquery.orders_id
	,mainSubquery.uploaded_orders_id
	,mainSubquery.uploaded_amended_orders_id
FROM PUBLIC.moves moves
JOIN (
	SELECT move_id
		,orders_id
		,uploaded_orders_id
		,uploaded_amended_orders_id
	FROM (
		SELECT downloadable_moves.id move_id
			,downloadable_moves.STATUS
			,downloadable_moves.orders_id
		FROM (
			SELECT moves.*
			FROM PUBLIC.moves moves
			JOIN (
				SELECT orders.id AS order_id
				FROM PUBLIC.orders orders
				JOIN (
					SELECT document_id
					FROM (
						SELECT document_id
							,count(document_id)
						FROM PUBLIC.user_uploads
						GROUP BY document_id
						) t
					WHERE count >= 1
					) doc_cnt_view ON orders.uploaded_orders_id = doc_cnt_view.document_id
				) order_uploaded_view ON moves.orders_id = order_uploaded_view.order_id
			) moves_by_uploaded_doc_cnt
		JOIN (
			SELECT m.*
			FROM (
				SELECT o.*
				FROM PUBLIC.orders o
				JOIN PUBLIC.duty_locations dl ON o.origin_duty_location_id = dl.id
					AND dl.provides_services_counseling = false
				) t
			JOIN PUBLIC.moves m ON t.id = m.orders_id
			) downloadable_moves ON moves_by_uploaded_doc_cnt.id = downloadable_moves.id
		) innerSubquery
	JOIN PUBLIC.orders orders ON orders.id = innerSubquery.orders_id
	) mainSubquery ON moves.id = mainSubquery.move_id;
```

### Backend
